test/server/guarddog_test_interlock.h

```diff
@@ -0,0 +1,37 @@
+#include "envoy/thread/thread.h"
+
+#include "source/server/guarddog_impl.h"
+
+#pragma once
+
+namespace Envoy {
+namespace Server {
+
+// Helps make tests using the GuardDog more robust by providing a way of
+// blocking in the test on GuardDog loop completion when
+// GuardDogImpl::forceCheckForTest() is called. If this interlock is not
+// provided, the default interlock is a no-op.
+class GuardDogTestInterlock : public GuardDogImpl::TestInterlockHook {
+public:
+  // GuardDogImpl::TestInterlockHook
+  void signalFromImpl() override {
+    waiting_for_signal_ = false;
+    impl_.notifyAll();
+  }
+
+  void waitFromTest(Thread::MutexBasicLockable& mutex) override
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex) {
+    ASSERT(!waiting_for_signal_);
+    waiting_for_signal_ = true;
```

Comment:
Note: I know we can probably use Mutex::Await or ThreadSynchronizer for this rather than a condvar, but this PR is not changing this code; it's just moving it.